clang/lib/Sema/SemaExpr.cpp

```diff
@@ -241,6 +241,19 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
                !checkAllowedSYCLInitializer(VD, /*CheckValueDependent =*/true))
         SYCLDiagIfDeviceCode(*Locs.begin(), diag::err_sycl_restrict)
             << Sema::KernelConstStaticVariable;
+    } else if (isa<FunctionDecl>(D)) {
+      // SYCL device function cannot be called from ESIMD context. However,
+      // there are some device function declarations that are shared between
+      // SYCL and ESIMD, e.g. spirv builtins. Those are reserved functions
+      // and we allow to call them from ESIMD context.
+      FunctionDecl *FDecl = cast<FunctionDecl>(D);
+      const IdentifierInfo *Id = FDecl->getIdentifier();
+      if ((getEmissionReason(FDecl) == Sema::DeviceDiagnosticReason::Sycl) &&
+          Id && !Id->isReservedName(/*doubleUnderscoreOnly=*/true)) {
```

Comment:
```suggestion
      if (Id && getEmissionReason(FDecl) == Sema::DeviceDiagnosticReason::Sycl &&
          !Id->isReservedName(/*doubleUnderscoreOnly=*/true)) {
```
The check for a `nullptr` will be faster than the function call.

Reply:
Yes, but I think we will almost never exit `if` block through the check for a `nullptr`.